### PR TITLE
feat: introduce SyncWaveHook callbacks which are invoked after applying each sync wave

### PR DESCRIPTION
**Pull request by jessesuen**
This is groundwork for fixing https://github.com/argoproj/argo-cd/issues/4669. It introduces a SyncWaveHook callback function, which is invoked after successfully applying each sync wave. Argo CD will leverage this by sleeping for a few seconds between each wave.

Also this fixes some logging issues after switching to the new logr.

### CODE REVIEW

File: pkg/sync/common/types.go [hunk 1]
```diff
@@ -29,6 +29,11 @@ type PermissionValidator func(un *unstructured.Unstructured, res *metav1.APIReso
 
 type SyncPhase string
 
+// SyncWaveHook is a callback function which will be invoked after each sync wave is successfully
+// applied during a sync operation. The callback indicates which phase and wave it had just
+// executed, and whether or not that wave was the final one.
+type SyncWaveHook func(phase SyncPhase, wave int, final bool) error
```
Comment (jessesuen) [inline pkg/sync/common/types.go:35]: Other name I considered was: SyncWaveCallback

File: pkg/sync/sync_context.go [hunk 2]
```diff
@@ -400,6 +410,15 @@ func (sc *syncContext) Sync() {
 
 	sc.log.WithValues("tasks", tasks).V(1).Info("Wet-run")
 	runState := sc.runTasks(tasks, false)
+
+	if sc.syncWaveHook != nil && runState != failed {
```
Comment (mayzhang2000) [inline pkg/sync/sync_context.go:414]: Here if `runState` is failed, `SyncWaveHook` is not called. I wonder if there is a use case that `SyncWaveHook `needs the flexibility to do different things based on the `runState`? Do we want the input parameter of SyncWaveHook also includes runState?

**Review by mayzhang2000 — APPROVED** [review]
LGTM

**Review by alexmt — APPROVED** [review]
LGTM

File: pkg/sync/sync_context.go [hunk 3]
```diff
@@ -400,6 +410,15 @@ func (sc *syncContext) Sync() {
...
+		err := sc.syncWaveHook(phase, wave, finalWave)
+		if err != nil {
+			sc.setOperationPhase(common.OperationFailed, fmt.Sprintf("SyncWaveHook failed: %v", err))
+			runState = failed
```
Comment (jessesuen) [inline pkg/sync/sync_context.go:418]: I just noticed the operation message will get clobbered by the one below. Will fix it before merging.

**Pull request merged**